files/__addonLocation__/tsconfig.json

```diff
@@ -0,0 +1,20 @@
+{
+  "extends": "@tsconfig/ember/tsconfig.json",
+  "compilerOptions": {
+    "paths": {
+      // No paths, no absolute imports, only node_modules imports allowed
+      // But fallback for type-overrides and such
+      "*": ["types/*"]
```

Comment:
do we want to allow this for the addon?

consuming projects wouldn't know about these types and could error as a result

Reply:
Not sure what you mean here?

I took it from your [comment here](https://github.com/embroider-build/embroider/pull/1099#discussion_r889573849) actually 😏

Reply:
hmm I'll investigate

Reply:
I think v2 addons should aim to avoid needing `paths` config at all. There's always `declare module` if absolutely necessary, but in general I don't think we want to make it easier for addon authors to write local type declarations that will break things when their consumers don't have them.

Reply:
it turns out this is needed if a v2 addon wants the cached-decorator types. 
the cached-decorator types go in the types directory, but we don''t want to provide those types, because they could conflict with the consumer's types for `@cached`

Reply:
What's special about `@cached` that makes it require a `paths` entry?

Reply:
Isn't `*: [types/*]` how you tell ts to check every file in that folder for type defs?

Reply:
1. `'*': ['types/*']` says "when any import fails, see if the mapping here produces the path to a module you can load instead, and if so, treat that as implicitly being the contents of the thing you were trying to look up"

2. `include: ['types']` says "consider everything under `types/` implicitly in scope as part of this project, whether it's imported or not" (roughly equivalent to specifying entrypoints for a bundler)

3. A `declare module` block _anywhere_ covered by a project's `include` will be taken into account, even without path mapping shenanigans.
 
Given (2) and (3), anything you're implicitly relying on `paths` for can be wrapped in `declare module` instead. You can still align your declarations on disk 1:1 with the module layout if you want, but `declare module` also gives you the flexibility to e.g. group together declarations for a package in a single `.d.ts` file if you want.

Reply:
excellent! thanks for explaining those things!